domain/machine/state/state.go

```diff
@@ -499,3 +499,89 @@ func (st *State) AllMachineNames(ctx context.Context) ([]machine.Name, error) {
+// TODO(cderici): use machineerrors.MachineNotFound on rebase after #17759
+// lands.
+func (st *State) MarkMachineForRemoval(ctx context.Context, mName machine.Name) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare query for getting the machine UUID.
+	machineNameParam := machineName{Name: mName}
+	markForRemovalWithUUID := machineMarkForRemoval{}
+	machineUUIDQuery := `SELECT uuid AS &machineMarkForRemoval.machine_uuid FROM machine WHERE name = $machineName.name`
+	machineUUIDStmt, err := st.Prepare(machineUUIDQuery, machineNameParam, markForRemovalWithUUID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare query for adding the machine to the machine_removals table.
+	markForRemovalWithUUID.Mark = true
+	markForRemovalUpdateQuery := `
+INSERT INTO machine_removals (machine_uuid, mark_for_removal)
+VALUES ($machineMarkForRemoval.*)`
+	markForRemovalStmt, err := st.Prepare(markForRemovalUpdateQuery, markForRemovalWithUUID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Query for the machine UUID.
+		err := tx.Query(ctx, machineUUIDStmt, machineNameParam).Get(&markForRemovalWithUUID)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(machineerrors.NotFound, "machine %q", mName)
+		}
+		if err != nil {
+			return errors.Annotatef(err, "querying UUID for machine %q", mName)
+		}
+
+		// Run query for updating the mark_for_removal column.
+		return tx.Query(ctx, markForRemovalStmt, markForRemovalWithUUID).Run()
+	})
+
+	return errors.Annotatef(err, "marking machine %q for removal", mName)
+}
+
+// AllMachineRemovals returns the UUIDs of all of the machines that need to be
+// removed but need provider-level cleanup.
+func (st *State) AllMachineRemovals(ctx context.Context) ([]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Prepare query for uuids of all machines marked for removal.
+	markForRemovalParam := machineMarkForRemoval{}
+	machinesMarkedForRemovalQuery := `SELECT &machineMarkForRemoval.* FROM machine_removals WHERE mark_for_removal = true`
+	machinesMarkedForRemovalStmt, err := st.Prepare(machinesMarkedForRemovalQuery, markForRemovalParam)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var results []machineMarkForRemoval
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Run query to get all machines marked for removal.
+		err := tx.Query(ctx, machinesMarkedForRemovalStmt).GetAll(&results)
+		// No errors if there's no machine marked for removal.
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		}
+		return err
+	})
+
+	if err != nil {
+		return nil, errors.Annotate(err, "querying all machines marked for removal")
+	}
+
+	// Transform the results ([]machineUUID) into a slice of machine UUIDs.
+	machineUUIDs := transform.Slice[machineMarkForRemoval, string](
```

Comment:
**suggestion:** Remove the `[machineMarkForRemoval, string]` it is useless (golang is able to infers type anyway)